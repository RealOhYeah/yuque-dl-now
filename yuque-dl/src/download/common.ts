import * as stream from 'node:stream'
import { promisify } from 'node:util'
import { createWriteStream } from 'node:fs'
import axios from 'axios'

import { genCommonOptions } from '../api'

interface IDownloadFileParams {
  fileUrl: string,
  savePath: string,
  token?: string
  key?: string,
  fileName: string
}

const finished = promisify(stream.finished)
export async function downloadFile(params: IDownloadFileParams) {
  const {fileUrl, savePath, token, key, fileName} = params
  return axios.get(fileUrl, {
    ...genCommonOptions({token, key}),
    responseType: 'stream'
  }).then(async response => {
    if (response.request?.path?.startsWith('/login')) {
      throw new Error(`"${fileName}" need token`)
    } else if (response.status === 200) {
      const writer = createWriteStream(savePath)
      response.data?.pipe(writer)
      return finished(writer)
        .then(() => ({
          fileUrl,
          savePath
        }))
    }
    throw new Error(`response status ${response.status}`)
  })
}
