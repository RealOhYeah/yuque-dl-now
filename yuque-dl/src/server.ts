import { createServer } from 'vitepress'
import { mkdir, writeFile, readFile, readdir, stat, access } from 'node:fs/promises'
import { join, resolve, sep } from 'node:path'

let restartPromise: Promise<void> | undefined

export async function runServer(root: string) {
  const rootPath = resolve(root)
  if (!await fileExists(rootPath)) {
    throw new Error('server root not found')
  }
  await createVitePressConfig(rootPath)
  const createDevServer = async () => {
    const server = await createServer(root, {}, async () => {
      if (!restartPromise) {
        restartPromise = (async () => {
          await server.close()
          await createDevServer()
        })().finally(() => {
          restartPromise = undefined
        })
      }

      return restartPromise
    })
    await server.listen()
    server.printUrls()
    // bindShortcuts(server, createDevServer)
  }

  await createDevServer().catch((err) => {
    console.error(err)
    process.exit(1)
  })
}

async function createVitePressConfig(root: string) {
  const bookName = root.split(sep).filter(Boolean).at(-1) || 'yuque-dl'
  const vitepressPath = join(root, '/.vitepress')
  await mkdir(vitepressPath, {recursive: true})
  const vitePressConfig = join(vitepressPath, 'config.ts')
  const sidebar = await createSidebarMulti(root)
  const config = `
  export default {
    title: "${bookName}",
    themeConfig: {
      search: {
        provider: 'local'
      },
      sidebar: ${JSON.stringify(sidebar)}
    },
    vite: {
      optimizeDeps: {
        include: []
      }
    },
    markdown: {
      html: false
    }
  }
  `
  await writeFile(vitePressConfig, config)
}

async function fileExists(filename: string) {
  try {
    await access(filename)
    return true
  } catch (err) {
    if (err.code === 'ENOENT') {
      return false
    } else {
      throw err
    }
  }
}

async function createSidebarMulti (path: string) {
  const data = [] as any
  const ignoreList = ['.vitepress', 'img', 'index.md', 'progress.json', 'attachments']
  let dirList = await readdir(path)
  dirList = dirList.filter(item => !ignoreList.includes(item))
  for (const n of dirList) {
    const dirPath = join(path, n)
    const statRes = await stat(dirPath)
    if (statRes.isDirectory()) {
      const isHasIndex = await fileExists(join(dirPath,'index.md'))
      const item: any = {
        text: n,
        collapsed: true,
        items: await createSideBarItems(path, n)
      }
      if (isHasIndex) item.link = `/${n}/`
      data.push(item)
    } else {
      data.push({
        text: n.slice(0, n.lastIndexOf('.')),
        link: `/${n}`
      })
    }
  }
  return data
}

// 尝试从一个md文件中读取标题，读取到第一个 ‘# 标题内容’ 的时候返回这一行
export async function getTitleFromFile (realFileName: string): Promise<string | undefined> {
  const isExist = await fileExists(realFileName)
  if (!isExist) {
    return undefined
  }
  const fileExtension = realFileName.substring(
    realFileName.lastIndexOf('.') + 1
  )
  if (fileExtension !== 'md' && fileExtension !== 'MD') {
    return undefined
  }
  // read contents of the file
  const data = await readFile(realFileName, { encoding: 'utf-8' })
  // split the contents by new line
  const lines = data.split(/\r?\n/)
  // return title
  for (const line of lines) {
    if (line.startsWith('# ')) {
      return line.substring(2)
    }
  }
  return undefined
}

async function createSideBarItems (
  targetPath: string,
  ...reset: string[]
) {
  const collapsed = false
  const node = await readdir(join(targetPath, ...reset))

  const result = []
  for (const fname of node) {
    const curPath = join(targetPath, ...reset, fname)
    const statRes = await stat(curPath)
    if (statRes.isDirectory()) {
      const items = await createSideBarItems(join(targetPath), ...reset, fname)
      if (items.length > 0) {
        const sidebarItem: any = {
          text: fname,
          items
        }
        // vitePress sidebar option collapsed
        sidebarItem.collapsed = collapsed
        const isHasIndex = await fileExists(join(curPath, 'index.md'))
        if (isHasIndex) {
          sidebarItem.link = '/' + [...reset, fname].map(decodeURIComponent).join('/') + '/'
        }
        result.push(sidebarItem)
      }
    } else {
      // is filed
      if (
        fname === 'index.md' ||
        /^-.*\.(md|MD)$/.test(fname) ||
        !fname.endsWith('.md')
      ) {
        continue
      }
      const fileName = fname.replace(/\.md$/, '')
      const item = {
        text: fileName,
        link: '/' + [...reset.map(decodeURIComponent), `${encodeURI(fileName)}.html`].join('/')
      }
      result.push(item)
    }
  }
  return result
}
