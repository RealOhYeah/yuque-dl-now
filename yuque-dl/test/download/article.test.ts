import path from 'node:path'
import { readdirSync, readFileSync } from 'node:fs'
import { vi, afterAll, afterEach, beforeAll, beforeEach, describe, expect, it } from 'vitest'
import { TestTools } from '../helpers/TestTools'
import { server } from '../mocks/server'
import { downloadArticle } from '../../src/download/article'

let testTools: TestTools

describe('downloadArticle', () => {
  beforeAll(() => {
    server.listen({ onUnhandledRequest: 'error' })
  })
  afterAll(() => server.close())

  beforeEach(() => {
    testTools = new TestTools()
  })

  afterEach(() => {
    testTools.cleanup()
    server.resetHandlers()
  })
  it('should work', async () => {
    const articleInfo = {
      savePath: testTools.cwd,
      saveFilePath: path.join(testTools.cwd, 'test.md'),
      host: 'https://www.yuque.com',
      ignoreImg: false,
      uuid: 'img_dir_uuid',
      articleTitle: 'downloadArticle Title',
      articleUrl: 'https://www.yuque.com/yuque/base1/one',
      itemUrl: 'one',
      bookId: 1111,
      imageServiceDomains: ['gxr404.com']
    }
    await downloadArticle({
      articleInfo,
      progressBar: {
        pause: vi.fn(),
        continue: vi.fn()
      } as any,
      options: {
        token: 'options token',
        key: 'options key'
      } as any
    })

    let doc1Data = readFileSync(articleInfo.saveFilePath).toString()
    doc1Data = doc1Data.replace(/\.\/img.*?-(\d{6})\./g, (match, random) => {
      return match.replace(random, '123456')
    })
    expect(doc1Data).toMatchSnapshot()
    const imgList = readdirSync(`${articleInfo.savePath}/img/${articleInfo.uuid}`)
    expect(readFileSync(`${articleInfo.savePath}/img/${articleInfo.uuid}/${imgList[0]}`).length).toBe(99892)
    expect(readFileSync(`${articleInfo.savePath}/img/${articleInfo.uuid}/${imgList[1]}`).length).toBe(81011)
  })

  it('sourcecode null should work', async () => {
    const articleInfo = {
      savePath: testTools.cwd,
      saveFilePath: path.join(testTools.cwd, 'test.md'),
      host: 'https://www.yuque.com',
      ignoreImg: false,
      uuid: 'img_dir_uuid',
      articleTitle: 'downloadArticle Title',
      articleUrl: 'https://www.yuque.com/yuque/base1/one',
      itemUrl: 'sourcecodeNull',
      bookId: 1111,
      imageServiceDomains: ['gxr404.com']
    }
    let isValidate = false
    try {
      await downloadArticle({
        articleInfo,
        progressBar: {
          pause: vi.fn(),
          continue: vi.fn()
        } as any,
        options: {
          token: 'options token',
          key: 'options key'
        } as any
      })
    } catch(e) {
      isValidate = true
      expect(e.message).toMatch( /download article Error: .*?, http status 200/g)
    }
    expect(isValidate).toBeTruthy()
  })


  it('board type', async () => {
    const articleInfo = {
      savePath: testTools.cwd,
      saveFilePath: path.join(testTools.cwd, 'test.md'),
      host: 'https://www.yuque.com',
      ignoreImg: false,
      uuid: 'img_dir_uuid',
      articleTitle: 'downloadArticle Title',
      articleUrl: 'https://www.yuque.com/yuque/base1/one',
      itemUrl: 'board',
      bookId: 1111,
      imageServiceDomains: ['gxr404.com']
    }
    let isValidate = false
    try {
      await downloadArticle({
        articleInfo,
        progressBar: {
          pause: vi.fn(),
          continue: vi.fn()
        } as any,
        options: {
          token: 'options token',
          key: 'options key'
        } as any
      })
    } catch (e) {
      expect(e.message).toBe('download article Error: 暂不支持“画板类型”的文档')
      isValidate = true
    }
    expect(isValidate).toBeTruthy()
  })

  it('table type', async () => {
    const articleInfo = {
      savePath: testTools.cwd,
      saveFilePath: path.join(testTools.cwd, 'test.md'),
      host: 'https://www.yuque.com',
      ignoreImg: false,
      uuid: 'img_dir_uuid',
      articleTitle: 'downloadArticle Title',
      articleUrl: 'https://www.yuque.com/yuque/base1/one',
      itemUrl: 'table',
      bookId: 1111,
      imageServiceDomains: ['gxr404.com']
    }
    let isValidate = false
    try {
      await downloadArticle({
        articleInfo,
        progressBar: {
          pause: vi.fn(),
          continue: vi.fn()
        } as any,
        options: {
          token: 'options token',
          key: 'options key'
        } as any
      })
    } catch (e) {
      expect(e.message).toBe('download article Error: 暂不支持“数据表类型”的文档')
      isValidate = true
    }
    expect(isValidate).toBeTruthy()
  })

  it('sheet type', async () => {
    const articleInfo = {
      savePath: testTools.cwd,
      saveFilePath: path.join(testTools.cwd, 'test.md'),
      host: 'https://www.yuque.com',
      ignoreImg: false,
      uuid: 'img_dir_uuid',
      articleTitle: 'downloadArticle Title',
      articleUrl: 'https://www.yuque.com/yuque/base1/one',
      itemUrl: 'sheet',
      bookId: 1111,
      imageServiceDomains: ['gxr404.com']
    }
    await downloadArticle({
      articleInfo,
      progressBar: {
        pause: vi.fn(),
        continue: vi.fn()
      } as any,
      options: {
        token: 'options token',
        key: 'options key'
      } as any
    })

    let doc1Data = readFileSync(articleInfo.saveFilePath).toString()
    doc1Data = doc1Data.replace(/\.\/img.*?-(\d{6})\./g, (match, random) => {
      return match.replace(random, '123456')
    })
    expect(doc1Data).toMatchSnapshot()
    const imgList = readdirSync(`${articleInfo.savePath}/img/${articleInfo.uuid}`)
    expect(readFileSync(`${articleInfo.savePath}/img/${articleInfo.uuid}/${imgList[0]}`).length).toBe(99892)
  })

  it('sheet type parse error', async () => {
    const articleInfo = {
      savePath: testTools.cwd,
      saveFilePath: path.join(testTools.cwd, 'test.md'),
      host: 'https://www.yuque.com',
      ignoreImg: false,
      uuid: 'img_dir_uuid',
      articleTitle: 'downloadArticle Title',
      articleUrl: 'https://www.yuque.com/yuque/base1/one',
      itemUrl: 'sheetError',
      bookId: 1111,
      imageServiceDomains: ['gxr404.com']
    }
    let isValidate = false
    try {
      await downloadArticle({
        articleInfo,
        progressBar: {
          pause: vi.fn(),
          continue: vi.fn()
        } as any,
        options: {
          token: 'options token',
          key: 'options key'
        } as any
      })
    } catch (e) {
      isValidate = true
      expect(e.message).toMatch(/download article Error: “表格类型”解析错误 SyntaxError: Unexpected token/)
    }
    expect(isValidate).toBeTruthy()
  })

  it('custom key token', async () => {
    const articleInfo = {
      savePath: testTools.cwd,
      saveFilePath: path.join(testTools.cwd, 'test.md'),
      host: 'https://www.yuque.com',
      ignoreImg: false,
      uuid: 'img_dir_uuid',
      articleTitle: 'downloadArticle Title',
      articleUrl: 'https://www.yuque.com/yuque/base1/one',
      itemUrl: 'tokenAndKey',
      bookId: 1111,
      imageServiceDomains: ['gxr404.com']
    }
    await downloadArticle({
      articleInfo,
      progressBar: {
        pause: vi.fn(),
        continue: vi.fn()
      } as any,
      options: {
        token: 'MyToken',
        key: 'MyKey'
      } as any
    })
    const docData = readFileSync(articleInfo.saveFilePath).toString()
    expect(docData).toMatchSnapshot()
  })

  it('uuid contains special characters', async () => {
    const articleInfo = {
      savePath: testTools.cwd,
      saveFilePath: path.join(testTools.cwd, 'test.md'),
      host: 'https://www.yuque.com',
      ignoreImg: false,
      uuid: 'img:dir/uuid',
      articleTitle: 'downloadArticle Title',
      articleUrl: 'https://www.yuque.com/api/docs/attachments',
      itemUrl: 'attachments',
      bookId: 1111,
      imageServiceDomains: ['gxr404.com']
    }
    await downloadArticle({
      articleInfo,
      progressBar: {
        pause: vi.fn(),
        continue: vi.fn()
      } as any,
      options: {
        token: 'MyToken',
        key: 'MyKey'
      } as any
    })
    let doc1Data = readFileSync(articleInfo.saveFilePath).toString()
    doc1Data = doc1Data.replace(/\.\/img.*?-(\d{6})\./g, (match, random) => {
      return match.replace(random, '123456')
    })
    expect(doc1Data).toMatchSnapshot()
  })
})
