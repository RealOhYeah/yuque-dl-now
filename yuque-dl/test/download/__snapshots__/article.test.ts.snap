// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`downloadArticle > custom key token 1`] = `
"# downloadArticle Title

MyKey=MyToken;

> 原文: <https://www.yuque.com/yuque/base1/one>"
`;

exports[`downloadArticle > sheet type 1`] = `
"# downloadArticle Title


## Sheet1

| |A | B | C | D | E | F | G | H | I|
|--- |--- |--- |--- |--- |--- |--- |--- |--- |---|
| 1 | 1 |  |  |  |  |  |  |  | |
| 2 |  |  |  | GSSSAPPAP |  |  |  |  | |
| 3 |  |  |  |  |  |  |  |  | |
| 4 |  | 12 |  |  |  |  |  |  | |
| 5 |  |  |  | 123 |  |  |  |  | |
| 6 |  |  |  |  |  |  |  |  | |
| 7 |  |  |  |  |  |  |  |  | |
| 8 |  |  |  |  |  |  |  |  | |
| 9 |  |  |  |  |  |  |  |  | |
| 10 |  |  |  |  |  |  |  |  | |
| 11 |  |  |  |  |  |  |  |  | |
| 12 |  |  |  |  |  |  |  |  | |
| 13 |  |  |  |  |  |  | 234 |  | |
| 14 |  |  |  |  |  |  |  |  | 32423|

## Sheet2

| |A | B | C | D | E | F | G | H|
|--- |--- |--- |--- |--- |--- |--- |--- |---|
| 1 | s |  |  | [x]  | 1 |  |  | |
| 2 |  |  |  | [baidu](https://www.baidu.com) | 13,21321 |  |  | |
| 3 |  | 34 |  | ![2x-00037-3212833155.png'](./img/img_dir_uuid/images-123456.jpeg) |  |  |  | |
| 4 |  | 34 |  | ![2x-00037-3212833155.png'](./img/img_dir_uuid/images-123456.jpeg) |  |  |  | |

## Sheet3



> 原文: <https://www.yuque.com/yuque/base1/one>"
`;

exports[`downloadArticle > should work 1`] = `
"# downloadArticle Title

# DOC1
![1.jpeg](./img/img_dir_uuid/1-123456.jpeg)
## SubTitle
![2.jpeg](./img/img_dir_uuid/2-123456.jpeg)

> 原文: <https://www.yuque.com/yuque/base1/one>"
`;

exports[`downloadArticle > uuid contains special characters 1`] = `
"# downloadArticle Title

# DOC1
![1.jpeg](./img/img_dir/uuid/1-123456.jpeg)
## SubTitle
[附件: test.pdf](./attachments/img_dir_uuid/test.pdf)

> 原文: <https://www.yuque.com/api/docs/attachments>"
`;
